### ISSUE TITLE
Determine the security properties of the new SPIRE Agent privileged API

### ISSUE THREAD

**Issue by evan2645** [spiffe/spire#2267]
Pursuant to https://github.com/spiffe/spire/issues/2246, we need a new privileged API for node local platform components to retrieve SVIDs on behalf of colocated workloads.

This is highly security sensitive, as currently it is not possible to ask the agent for an SVID that does not belong to you. Gaining access to this API could have the same effect on SPIRE as a full node compromise. The security properties of this API need to be thought out carefully.

For this issue, we should get consensus on exactly what properties we are looking for, and what mechanisms we will use to achieve them. Specifically, I think we need answers to:
* How do we control physical access (i.e. reachability) to this API?
* How do we authenticate callers of this API?
* Which callers are authorized and how?
* What limitations are necessary (if any) on the SVIDs or data that we will return on this API? (e.g. admin SVIDs are not available via this API)

Once we have consensus on the above points, and https://github.com/spiffe/spire-api-sdk/issues/4 is solved, we will open further issues to actually implement

**Comment by evan2645** [comment]
> How do we control physical access (i.e. reachability) to this API?

My feeling here is UDS only, non-workload-api socket, disabled by default

> How do we authenticate callers of this API?

I came up with two ideas

**Option 1:** callers are authenticated via SPIFFE mTLS. They fetch an SVID from the workload API, and use it to connect to the privileged API. Pro: consistent, idiomatic, straightforward implementation. Con: Requires the agent to process more attacker-controlled data than is strictly necessary (i.e. x.509 and tls now part of agent attack surface)

**Option 2:** callers are attested using standard workload attestation. Their identity is derived from the result. Pro: easy on the caller, no new attack surface facing unauthenticated callers. Con: trickier implementation, need to handle multiple identity match case

> Which callers are authorized and how?

Since the expectation is that these delegated platform components (our callers) will be spread around the infra far and wide, I think an admin SVID here is a bad idea. @amartinezfayo has mentioned the relation to https://github.com/spiffe/spire/issues/1975 and I agree... however, I think that is not happening soon, and rather than putting that work in scope of this work, we should put this work in scope of that work (i.e. find a short-term solution for this, and make sure it is considered when working on #1975)

With that, I think simply defining the SPIFFE ID of the authorized callers as part of spire-agent config makes sense.

> What limitations are necessary (if any) on the SVIDs or data that we will return on this API? (e.g. admin SVIDs are not available via this API)

I'm not sure what all can be done here. Not surfacing admin SVIDs here seems like a reasonable thing to do, but its value is dependent on how powerful we think our caller is. For example, if we assume that our caller has or can get root, then they could get admin SVIDs anyway from memory.

Are there any extra measures that come to mind here?

**Comment by azdagron** [comment]
> My feeling here is UDS only, non-workload-api socket, disabled by default
I agree. That sounds the most prudent. We already have a non-default non-workload-api socket called the "admin" socket. It hosts the debug API. I'm not sure if it would be appropriate to host this API there, but having three sockets feels like a little much. We could authorize access to the two APIs differently, of course, but exposes this socket would imply the caller has access to the Debug API since it is currently unauthenticated (physical access to the socket is the authorization bar)

> I came up with two ideas
I think I gravitate towards the second option, but not strongly so.

> With that, I think simply defining the SPIFFE ID of the authorized callers as part of spire-agent config makes sense.
I think that sound pretty straight-forward.

> Are there any extra measures that come to mind here?
I think filtering downstream entries in addition to admin entries would be prudent, at least in this first release. We can optionally open the valve later if we have a solid use case.

**Comment by evan2645** [comment]
This has been done. We went with UDS-only and used the admin socket... option two for leaning on workload attestation rather than explicit authentication. Authorized callers are configured by file. Downstream and admin entries are not available over this API.

In the course of implementation, we also realized that it is necessary to authorize the caller for every RPC issued, though authentication is done on the channel (workload attestation). To address this, we cache the selectors from attestation, and re-process the authorization logic for every RPC to ensure that the caller is still registered for the authorized identity.
